datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -875,6 +890,44 @@ public void cleanupOldUnusedTasks(Map<String, Set<DatastreamTask>> previousAssig
     unusedTasks.forEach(t -> deleteConnectorTask(t.getConnectorType(), t.getDatastreamTaskName()));
   }
 
+  /**
+   * Remove orphan connector task nodes which are not assigned to any instance (live or pause).
+   *
+   * NOTE: this should only be called after the valid tasks have been
+   * reassigned or become safe to discard per strategy requirement.
+   * This can be a costly operation. So, it should be call once the leader get elected and has finished
+   * the assignment and cleaned up dead Tasks. Ideally, we should not find anything in this check to clean up.
```

Comment:
Just a clarifying question: Any reason this shouldn't be done as part of cleaning up dead tasks in the original LEADER_DO_ASSIGNMENT event? Might be a good idea to add clear documentation around why this should only be performed afterwards.

Reply:
Updated the comment as well. get all children of /cluster/connector is a very costly operation. So, it should not be done very frequently.